python/ray/tests/test_state_api.py

```diff
@@ -2068,6 +2069,237 @@ def ready(self):
     assert "actor_id" in result.output
 
 
+def _try_state_query(api_func, q):
+    """Utility functions for rate limit related e2e tests below"""
+    try:
+        api_func()
+    except RayStateApiException as e:
+        # Other exceptions will be thrown
+        if "Max number of in-progress requests" in str(e):
+            q.put(1)
+        else:
+            q.put(e)
+    except Exception as e:
+        q.put(e)
+    else:
+        q.put(0)
+
+
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Lambda test functions could not be pickled on Windows",
+)
+def test_state_api_rate_limit_with_failure(monkeypatch, shutdown_only):
+    import queue
+    import multiprocessing as mp
+    import os
+    import signal
+
+    # Set environment
+    with monkeypatch.context() as m:
+        m.setenv("RAY_STATE_SERVER_MAX_HTTP_REQUEST", "3")
+        m.setenv(
+            "RAY_testing_asio_delay_us",
+            (
+                "NodeManagerService.grpc_server.GetTasksInfo=10000000:10000000,"
+                "NodeInfoGcsService.grpc_server.GetAllNodeInfo=1000000:1000000,"
+                "ActorInfoGcsService.grpc_server.GetAllActorInfo=10000000:10000000"
+            ),
+        )
+
+        # Set up scripts
+        ray.init()
+
+        @ray.remote
+        def f():
+            import time
+
+            time.sleep(30)
+
+        @ray.remote
+        class Actor:
+            pass
+
+        task = f.remote()  # noqa
+        actor = Actor.remote()  # noqa
+        actor_runtime_env = Actor.options(  # noqa
+            runtime_env={"pip": ["requests"]}
+        ).remote()
+        pg = ray.util.placement_group(bundles=[{"CPU": 1}])  # noqa
+
+        _objs = [ray.put(x) for x in range(10)]  # noqa
+
+        # list_objects() will wait for raylets to be registered, which
+        # means `list_tasks` will also see the registered raylets.
+        wait_for_condition(lambda: len(list_objects()) > 0)
+
+        # Running 3 slow apis to exhaust the limits
+        procs = [
+            mp.Process(target=lambda: print(list_tasks()), args=()),
+            mp.Process(target=lambda: print(list_nodes()), args=()),
+            mp.Process(target=lambda: print(list_actors()), args=()),
+        ]
+
+        [p.start() for p in procs]
+
+        # Running another 1 should return error
+        with pytest.raises(RayStateApiException):
```

Comment:
Can you do the match="" to make sure this is rate limiting error?

Reply:
Ah - i was doing so. Removed it for debugging but forgot to add it back. Good catch.

Reply:
Feel free to add a new exception. We are anyway planning to do that I think (ray.state.exceptions.RateLimitError)

Reply:
I was more thinking of an error code since I don't think exceptions could be auto serialized across server/clients? Like at the client side we will still need to check some sort of flags/field/error message to know what kind of exception it is.